src/test/groovy/com/gradle/BaseInitScriptTest.groovy

```diff
@@ -281,4 +295,96 @@ abstract class BaseInitScriptTest extends Specification {
             return "Gradle " + gradleVersion.version
         }
     }
+
+    static TestDvPluginVersion dvPlugin(DvPluginId id, String version, boolean deprecated = false) {
+        return new TestDvPluginVersion(id, version, deprecated)
+    }
+
+    static enum DvPluginId {
+        DEVELOCITY('com.gradle.develocity'),
+        GRADLE_ENTERPRISE('com.gradle.enterprise'),
+        BUILD_SCAN('com.gradle.build-scan');
+        final String id;
+
+        DvPluginId(String id) {
+            this.id = id
+        }
+    }
+
+    static final class TestDvPluginVersion {
+        final DvPluginId pluginId
+        final String version
+        final boolean deprecated
+
+        TestDvPluginVersion(DvPluginId pluginId, String version, boolean deprecated) {
+            this.pluginId = pluginId
+            this.version = version
+            this.deprecated = deprecated
+        }
+
+        String getId() {
+            return pluginId.id
+        }
+
+        boolean isCompatibleWith(TestGradleVersion gradleVersion) {
+            switch (pluginId) {
+                case DEVELOCITY:
+                    return GRADLE_5 <= gradleVersion.gradleVersion
+                case GRADLE_ENTERPRISE:
+                    return GRADLE_6 <= gradleVersion.gradleVersion
+                case BUILD_SCAN:
+                    if (version == '1.16') {
+                        // Only plugin v1.16 works with Gradle < 5
+                        return gradleVersion.gradleVersion < GRADLE_5
+                    } else {
+                        // Build-scan plugin 2+ only works with Gradle 5 (enterprise is for Gradle 6+)
+                        return GRADLE_5 <= gradleVersion.gradleVersion
+                            && gradleVersion.gradleVersion < GRADLE_6
+                    }
+            }
+        }
+
+        boolean isCompatibleWithConfigurationCache() {
+            // Only DV & GE plugins 3.16+ support configuration-cache
+            return pluginId != BUILD_SCAN
+                && GradleVersion.version(version) >= GradleVersion.version('3.16')
+        }
+
+        String getConfigBlock(URI serverUri) {
+            switch (pluginId) {
+                case DEVELOCITY:
+                    return """
+                        develocity {
+                            server = '$serverUri'
+                        }   
+                    """
+                case GRADLE_ENTERPRISE:
+                    return """
+                        gradleEnterprise {
+                            server = '$serverUri'
+                            buildScan { publishAlways() }
+                        }
+                    """
+                case BUILD_SCAN:
+                    return """
+                        buildScan {
+                            server = '$serverUri'
+                            publishAlways()
+                        }
+                    """
+            }
+        }
+
+        String getCompatibleCCUDVersion() {
```

Comment:
I had a comment here about CCUD compatibility, but we began discussing this in an internal Slack thread.